File: backend/src/Services/DutyRules.php
<?php

declare(strict_types=1);

namespace App\Services;

final class DutyRules
{
    public const CAPACITY = 3;

    public function __construct(private readonly string $timezone = 'Europe/Berlin')
    {
    }

    public function isWeekend(string $date): bool
    {
        $day = new \DateTimeImmutable($date, new \DateTimeZone($this->timezone));
        return in_array((int) $day->format('N'), [6, 7], true);
    }

    public function isWithinUpcomingWindow(string $date, ?\DateTimeImmutable $now = null): bool
    {
        $now ??= new \DateTimeImmutable('now', new \DateTimeZone($this->timezone));
        $today = $now->setTime(0, 0);
        $target = new \DateTimeImmutable($date, new \DateTimeZone($this->timezone));
        $last = $today->modify('+13 days');
        return $target >= $today && $target <= $last;
    }

    public function canBook(string $date, ?\DateTimeImmutable $now = null): bool
    {
        return !$this->isWeekend($date) && $this->isWithinUpcomingWindow($date, $now);
    }

    public function canCancelRegularly(string $date, ?\DateTimeImmutable $now = null): bool
    {
        $now ??= new \DateTimeImmutable('now', new \DateTimeZone($this->timezone));
        $dutyStart = new \DateTimeImmutable($date . ' 00:00:00', new \DateTimeZone($this->timezone));
        return ($dutyStart->getTimestamp() - $now->getTimestamp()) >= 48 * 3600;
    }

    public function canReportSick(string $date, ?\DateTimeImmutable $now = null): bool
    {
        $now ??= new \DateTimeImmutable('now', new \DateTimeZone($this->timezone));
        $today = $now->setTime(0, 0);
        $target = new \DateTimeImmutable($date, new \DateTimeZone($this->timezone));
        return $target >= $today && !$this->canCancelRegularly($date, $now);
    }
}

File: lib/utils/duty_rules.dart
class DutyRules {
  DutyRules._();

  static const capacity = 3;

  static bool isWeekend(DateTime day) =>
      day.weekday == DateTime.saturday || day.weekday == DateTime.sunday;

  static DateTime onlyDate(DateTime value) => DateTime(value.year, value.month, value.day);

  static bool isWithinUpcomingWindow(DateTime now, DateTime day) {
    final today = onlyDate(now);
    final target = onlyDate(day);
    final lastAllowed = today.add(const Duration(days: 13));
    return !target.isBefore(today) && !target.isAfter(lastAllowed);
  }

  static bool canBook(DateTime now, DateTime day) =>
      !isWeekend(day) && isWithinUpcomingWindow(now, day);

  static bool canCancelRegularly(DateTime now, DateTime day) {
    final dutyStart = DateTime(day.year, day.month, day.day);
    return dutyStart.difference(now).inMinutes >= 48 * 60;
  }

  static bool canReportSick(DateTime now, DateTime day) {
    final today = onlyDate(now);
    final target = onlyDate(day);
    if (target.isBefore(today)) return false;
    return !canCancelRegularly(now, day);
  }
}
